### PR TITLE
Further speedup isotovideo shutdown by 1-2s

### PR DESCRIPTION
**Pull request by okurz**
As a followup to 454c5dc0, same as for other subprocesses that we spawn we can
also use a flag to reduce the sleeptime during the shutdown of the "driver"
process, i.e. "backend".

I profiled with `(cd t/pool && perl -d:NYTProf ../../isotovideo -d)` to find
out the biggest time consumer during a short "start+stop" cycle of isotovideo.
This brought me back to the backend process which I thought I had covered in
454c5dc0, maybe forgot, maybe reverted for a good reason that we will learn
about later ;)

For a statistical proof I conducted runtime measurements.
Using '(cd t/pool && avgtime -r=5 -p ../../isotovideo -d)' without any
modifications I get:

Median time    : 6657.16
Avg time       : 6824.06
Std dev.       : 402.794
Minimum        : 6587.82
Maximum        : 7627.35

with `, sleeptime_during_kill => .1` on the process in driver::start:

Median time    : 5591.99
Avg time       : 5592.33
Std dev.       : 8.85996
Minimum        : 5580.75
Maximum        : 5605.03

so a significant improvement of 1.3+-0.4s which should be proof enough for the
change within this pull request.

Further I took the chance to test further combinations and alternatives, e.g.
just https://github.com/mudler/Mojo-IOLoop-ReadWriteProcess/pull/6 :

Median time    : 5203.46
Avg time       : 5353.43
Std dev.       : 212.99
Minimum        : 5141.65
Maximum        : 5639.83

showing another, independant improvement.

with https://github.com/mudler/Mojo-IOLoop-ReadWriteProcess/pull/6 and
additionally `, sleeptime_during_kill => .1`:

Median time    : 4272.04
Avg time       : 4430.27
Std dev.       : 251.437
Minimum        : 4190.39
Maximum        : 4829.96

We see that the positive effects accumulate.

with …pull/6 and `, sleeptime_during_kill => .01`:

Median time    : 4102.76
Avg time       : 4355.76
Std dev.       : 400.075
Minimum        : 4028.73
Maximum        : 5065.25

showing that no significant further improvement is visible.

with …pull/6 and `, sleeptime_during_kill => .01` and `, kill_sleeptime => .01`:

Median time    : 4538.72
Avg time       : 4446.54
Std dev.       : 187.31
Minimum        : 4072.02
Maximum        : 4546.59

also no significant further improvement.

For the "null" backend, n=50, with …pull/6 and `, sleeptime_during_kill => .01`
and `, kill_sleeptime => .01`:

Median time    : 1529.18
Avg time       : 1366.56
Std dev.       : 551.558
Minimum        : 626.798
Maximum        : 2773.82

"null" backend, n=50, with …pull/6 and `, sleeptime_during_kill => .01`:

Median time    : 1534.52
Avg time       : 1378.34
Std dev.       : 599.227
Minimum        : 687.546
Maximum        : 2901.23

"null" backend, n=50, with …pull/6 and `, sleeptime_during_kill => .1`:

Median time    : 1530.83
Avg time       : 1326.06
Std dev.       : 569.442
Minimum        : 689.181
Maximum        : 2558.02

"null" backend, n=50, with …pull/6:

Median time    : 2426.69
Avg time       : 2195.34
Std dev.       : 602.505
Minimum        : 700.167
Maximum        : 3625.74

"null" backend, n=50:

Median time    : 2518.98
Avg time       : 2411.55
Std dev.       : 568.33
Minimum        : 1519.95
Maximum        : 3631.11

Related progress issue: https://progress.opensuse.org/issues/58379

### CODE REVIEW

**Comment by okurz** [comment]
ok, should add a test using the null backend actively. Or we just switch t/14-isotovideo.t to using "null".

**Comment by okurz** [comment]
The big coverage drop is surprising to me. Shouldn't the fullstack test cover this?

File: t/14-isotovideo.t [hunk 1]
```diff
@@ -14,9 +14,12 @@ my $pool_dir     = "$toplevel_dir/t/pool";
...
     my (%args) = @_;
-    $args{opts}      //= '';
-    $args{exit_code} //= 1;
-    system("perl $toplevel_dir/isotovideo -d $args{opts} 2>&1 | tee autoinst-log.txt");
+    $args{default_opts} //= 'backend=null';
+    $args{opts}         //= '';
+    $args{exit_code}    //= 1;
+    my $cmd = "perl $toplevel_dir/isotovideo -d $args{default_opts} $args{opts} 2>&1 | tee autoinst-log.txt";
+    note("Starting isotovideo with: $cmd");
+    system($cmd);
```
Comment (kalikiana) [inline t/14-isotovideo.t]: Maybe merge `note`/`system` here a la `is(system(...`?

File: bmwqemu.pm [hunk 2]
```diff
@@ -155,7 +155,7 @@ sub _check_publish_vars {
         for my $type (qw(STORE PUBLISH FORCE_PUBLISH)) {
             my $name = $type . "_HDD_$i";
             next unless my $out = $vars{$name};
-            die "HDD_$i also specified in $name. This is not supported" if grep { $_ eq $out } @hdds;
+            die "HDD_$i also specified in $name. This is not supported" if grep { $_ && $_ eq $out } @hdds;
```
Comment (kalikiana) [inline bmwqemu.pm]: Is this for the null backend?

File: backend/driver.pm [hunk 3]
```diff
@@ -98,7 +98,7 @@ sub start {
 
             $self->{backend}->run(fileno($process->channel_in), fileno($process->channel_out));
             _exit(0);
-    })->blocking_stop(1)->separate_err(0)->subreaper(1)->start;
+    }, sleeptime_during_kill => .1)->blocking_stop(1)->separate_err(0)->subreaper(1)->start;
```
Comment (kalikiana) [inline backend/driver.pm]: Interesting find. This is [not even documented](https://metacpan.org/pod/Mojo::IOLoop::ReadWriteProcess). I guess you spotted its use in Mojo.

File: t/14-isotovideo.t (hunk 1 above)
Comment (okurz) [reply t/14-isotovideo.t]: I don't understand what you mean. "note" is there to just print some debug message on `prove -v`, `system` is executing something. There is nothing to compare `is` against here.

File: bmwqemu.pm (hunk 2 above)
Comment (okurz) [reply bmwqemu.pm]: That's independant of the backend, part of commit https://github.com/os-autoinst/os-autoinst/pull/1360/commits/6b6182d6a807c01c2d273e735b3e6eb127c1bc02 which can happen whenever this method is called without initializing the full stack in before but not only in unit tests but also when calling `isotovideo` locally under certain circumstances.

File: backend/driver.pm (hunk 3 above)
Comment (okurz) [reply backend/driver.pm]: yes, unfortunately. Only https://metacpan.org/pod/Mojo::IOLoop::ReadWriteProcess#kill_sleeptime is documented.

File: bmwqemu.pm (hunk 2 above)
Comment (kalikiana) [reply bmwqemu.pm]: Aha. Thanks for clarifying!

File: t/14-isotovideo.t (hunk 1 above)
Comment (kalikiana) [reply t/14-isotovideo.t]: The next line already does it, and checks that the exit code is `0`. That´s why I thought of it.

File: backend/null.pm [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package backend::null;
```
Comment (Martchus) [inline backend/null.pm]: Maybe add a comment that this backend is for internal testing?

Comment (okurz) [reply backend/null.pm]: Yeah I think we don't have a consistent list of backends extracted into documentation. But actually I would not say that the backend is meant to be used "only for internal testing". It can very well be used for "external testing", e.g. run tests of os-autoinst-distri-opensuse against that.
